plugin/core/promise.py

```diff
@@ -0,0 +1,183 @@
+from .typing import Any, Callable, List
+import functools
+import threading
+
+ResolveFunc = Callable[..., None]  # Optional argument not supported in Callable so using "..."
+FullfillFunc = Callable[[ResolveFunc], None]
+ThenFunc = Callable[[Any], Any]
+
+
+class Promise:
```

Comment:
This promise class should be generic in a single template type `T`. Just like we have `Callable[[S], T]`, we should be able to write `Promise[T]`. Given a `Promise[S]`, the `.then` continuation should take a function `S -> T` and return a `Promise[T]`. I hope it's possible to do this :)

Reply:
Furthermore if `.then` gets a function `S -> Promise[T]`, then instead of a `Promise[Promise[T]]`, it should collapse to a `Promise[T]` (the monadic join operation).